_posts/2020-09-02-XHP-namespaces-and-syntax.md

```diff
@@ -0,0 +1,221 @@
+  `$xhp->getChildrenOfType<...>()`, `$xhp->getFirstChildOfType<...>()`, etc.
+- Most deprecated features (e.g. attribute clobbering, `forceAttribute`) are not
+  handled by the automated migrations, we recommend replacing them with
+  equivalent or similar non-deprecated features.
+
+# Details
+
+## Language changes
+
+A [more formal](https://gist.github.com/fredemmott/de7eeb91765b0c98fafac9c591263aca) description of these changes is also available.
+
+### Syntax
+
+Previously, XHP classes were declared with a leading colon, e.g.
+
+```Hack
+class :foo:bar {}
+```
+
+XHP classes are now declared with the following syntax:
+```Hack
+namespace foo;
+
+xhp class bar {}
+```
+
+This new syntax avoids ambiguities that would otherwise be introduced by the addition of namespace support.
+
+`xhp class foo:bar {}` is also permitted - it is intended to be used:
+- by existing codebases, as a migration aid
+- as a stylistic preference by codebases that choose to minimize use of namespaces
+
+### Naming
+
+Previously, XHP class names were 'mangled' into a form that meets the usual restrictions for Hack class names; for example, `class :foo:bar-baz {}` actually defined the class `xhp_foo__bar_baz`.
+
+XHP class names are no longer mangled: `xhp class foo` defines the class `foo`, and `xhp class foo:bar` defines the class `foo\bar`; this may cause problems if you are storing or otherwise interacting with `classname<T>` of XHP classes, or storing serialized instances of XHP classes.
+
+This also means that XHP class names must now be valid Hack XHP class names; the most significant effect is that the `-` character is no longer permitted in XHP class names.
+
+### Colons in XHP class types
+
+Colons are now a namespace separator (largely equivalent to `\`), but are only permitted in some contexts:
+- in XHP-specific language constructs, such as `xhp class foo:bar`, `<foo:bar>`
+- when fully-qualified, in other contexts where qualified names are permitted - for example, `function foo(:bar:baz $_)` is permitted, but `function foo(bar:baz $_)` is not permitted
+
+`\` is not permitted in XHP constructor calls (e.g. `<foo\bar>` is an error), or XHP class declaration names (e.g. `xhp class foo\bar`). It is permitted in parent class references, e.g. `xhp class foo extends bar\baz {}`.
+
+Like `\`, `:` can be used to indicate that a name is fully qualified; for example, `foo\bar` and `foo:bar` both refer to `namespace\foo\bar` (unless there is a relevant `use` statement), and `\foo\bar` and `:foo:bar` both always refer to `\foo\bar`. Similarly, `<foo:bar />` is a reference to `namespace\foo\bar`, and `<:foo:bar />` is now permitted as a reference to `\foo\bar`.
+
+There are several situations where both `\` and `:` are permitted; we recommend that codebases decide on a consistent style; for example, code bases may wish to ban:
+- using `:` for anything that is not an XHP class
+- using `\` for any XHP class
```

Comment:
Not sure if this is a good recommendation since `:` is not allowed in most contexts (unless maybe it's fully qualified?)